funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/task_queue_subscriber.py

```diff
@@ -0,0 +1,401 @@
+        self.setup_exchange(self.exchange)
+
+    def add_on_channel_close_callback(self):
+        """This method tells pika to call the on_channel_closed method if
+        RabbitMQ unexpectedly closes the channel.
+
+        """
+        logger.info("Adding channel close callback")
+        self._channel.add_on_close_callback(self._on_channel_closed)
+
+    def _on_channel_closed(self, channel, exception):
+        """Invoked by pika when RabbitMQ unexpectedly closes the channel.
+        Channels are usually closed if you attempt to do something that
+        violates the protocol, such as re-declare an exchange or queue with
+        different parameters. In this case, we'll close the connection
+        to shutdown the object.
+
+        :param pika.channel.Channel: The closed channel
+        :param Exception: exception from pika
+        """
+        logger.warning(f"Channel:{channel} was closed: ({exception}")
+        try:
+            logger.warning("Raising exception:")
+            raise exception
+        except pika.exceptions.ChannelClosedByBroker:
+            if "exclusive use" in exception.reply_text:
+                logger.exception(
+                    "Channel closed by RabbitMQ broker due to exclusive "
+                    "ownership by an active endpoint"
+                )
+                logger.warning("Channel will close without connection retry")
+                self._closing = True
+        except Exception as e:
+            logger.exception(
+                f"Channel closed with code:{e.reply_code}, error:{e.reply_text}"
+            )
+
+        self._connection.close()
+
+    def setup_exchange(self, exchange_name):
+        """Setup the exchange on RabbitMQ by invoking the Exchange.Declare RPC
+        command. When it is complete, the on_exchange_declareok method will
+        be invoked by pika.
+
+        :param str|unicode exchange_name: The name of the exchange to declare
+        """
+        logger.info(f"Declaring exchange {exchange_name}")
+        self._channel.exchange_declare(
+            exchange=exchange_name,
+            exchange_type=self.exchange_type,
+            callback=self._on_exchange_declareok,
+        )
+
+    def _on_exchange_declareok(self, unused_frame):
+        """Invoked by pika when RabbitMQ has finished the Exchange.Declare RPC
+        command.
+
+        :param pika.Frame.Method unused_frame: Exchange.DeclareOk response frame
+
+        """
+        logger.info("Exchange declared")
+        self.setup_queue(self.queue_name)
+
+    def setup_queue(self, queue_name):
+        """Setup the queue on RabbitMQ by invoking the Queue.Declare RPC
+        command. When it is complete, the on_queue_declareok method will
+        be invoked by pika.
+
+        :param str|unicode queue_name: The name of the queue to declare.
+
+        """
+        logger.info("Declaring queue %s", queue_name)
+        self._channel.queue_declare(queue_name, callback=self._on_queue_declareok)
```

Comment:
```suggestion
        logger.info("Declaring queue %s", self.queue_name)
        self._channel.queue_declare(self.queue_name, callback=self._on_queue_declareok)
```

This class, like the other one, can have dozens of physical lines removed if we just remove extra layers of indirection. I'm not going to enumerate all of them here, but this needs a pass to simplify.